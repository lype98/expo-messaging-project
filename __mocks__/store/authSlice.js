import { createSlice } from '@reduxjs/toolkit';

const authSlice = createSlice({
  name: 'auth',
  initialState: {
    uid: 'gR9pqEb0bpehle4GHbEEUWiu5U43',
    email: 'a@a.com',
    username: 'A boy!',
  },
  reducers: {
    login(state, action) {
      state.uid = action.payload.uid;
      state.email = action.payload.email;
      state.username = action.payload.username;
    },
    logOut(state) {
      for (let key in state) {
        state[key] = '';
      }
    },
  },
});

export const authActions = authSlice.actions;

export default authSlice;
