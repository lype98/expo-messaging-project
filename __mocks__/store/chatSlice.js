import { createSlice } from '@reduxjs/toolkit';

const chatSlice = createSlice({
  name: 'chat',
  initialState: {
    currentRoomID: '9hr8DpuV2YOcv73D89Hc',
    currentRoomType: 'Group',
    currentRoomUsers: [
      {
        id: 'gR9pqEb0bpehle4GHbEEUWiu5U43',
        displayName: 'A boy!',
      },
      {
        id: 'IZ7rvcBzKbOqyuaopfWmDi1y6Bv2',
        displayName: 'BBoi',
      },
      {
        id: 'kBp1Bvh2jIdPpnO8TfSiV59ZqtB2',
        displayName: 'C se√±or',
      },
    ],
    refreshing: false,
  },
  reducers: {
    enterRoom(state, action) {
      state.currentRoomID = action.payload.roomID;
      state.currentRoomType = action.payload.chatType;
      state.currentRoomUsers = action.payload.users;
    },
    refresh(state) {
      state.refreshing = !state.refreshing;
    },
  },
});

export const chatActions = chatSlice.actions;

export default chatSlice;
