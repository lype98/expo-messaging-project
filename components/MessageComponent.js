import { StyleSheet, Text, View, Image } from 'react-native';
import { useSelector } from 'react-redux';
import { format } from 'date-fns';
import { Video } from 'expo-av';

import avatar from '../assets/avatar1.jpg';

const MessageComponent = ({ messageObj, otherUser, chatType }) => {
  const currentUser = useSelector((state) => state.auth);
  const currentRoomUsers = useSelector((state) => state.chat.currentRoomUsers);
  const formattedTime = format(new Date(messageObj.createdAt.seconds * 1000), 'p');

  let senderName;
  let avatarJSX;
  let messageStyle = {};
  const isSystem = messageObj.uid === 'system';

  if (isSystem) {
    senderName = 'system';
    messageStyle = {
      container: 'sysContainer',
      title: '',
      message: 'sysMessage',
      msgText: 'sysMsgText',
      time: '',
    };
  } else if (messageObj.uid === currentUser.uid) {
    senderName = currentUser.username;
    messageStyle = {
      container: 'myContainer',
      title: 'myMsgTitle',
      message: 'myMessage',
      msgText: 'myMsgText',
      time: 'myTime',
    };
  } else {
    if (chatType === 'DM') senderName = otherUser;
    if (chatType === 'Group') {
      for (const user of currentRoomUsers) {
        if (user.id === messageObj.uid) senderName = user.displayName;
      }
    }
    messageStyle = {
      container: 'theirContainer',
      title: 'theirMsgTitle',
      message: 'theirMessage',
      msgText: 'theirMsgText',
      time: 'theirTime',
    };
  }

  if (messageStyle.message === 'theirMessage') {
    avatarJSX = <Image style={styles.avatar} source={avatar} />;
  }

  const DisplayMsgContent = () => {
    if (!messageObj.mediaType) {
      return <Text style={styles[`${messageStyle.msgText}`]}>{messageObj.message}</Text>;
    }
    if (messageObj.mediaType.substring(5, 0) === 'image') {
      return (
        <View style={{}}>
          <Image style={{}} source={{ uri: messageObj.src, width: 100, height: 100 }} />
        </View>
      );
    }
    if (messageObj.mediaType.substring(5, 0) === 'video') {
      return (
        <View style={{}}>
          <Video style={{ width: 200, height: 200 }} source={{ uri: messageObj.src }} useNativeControls isLooping resizeMode="contain" />
        </View>
      );
    } else {
      return <Text style={styles[`${messageStyle.msgText}`]}>File not supported</Text>;
    }
  };

  return (
    <View style={styles[`${messageStyle.container}`]}>
      {avatarJSX}
      <View>
        {!isSystem && (
          <Text testID="msgTitle" style={styles[`${messageStyle.title}`]}>
            {senderName}
          </Text>
        )}
        <View style={styles[`${messageStyle.message}`]}>
          <DisplayMsgContent />
          {!isSystem && <Text style={styles[`${messageStyle.time}`]}>{formattedTime}</Text>}
        </View>
      </View>
    </View>
  );
};

export default MessageComponent;

const styles = StyleSheet.create({
  myContainer: {
    flexDirection: 'row',
    justifyContent: 'flex-end',
  },
  theirContainer: {
    flexDirection: 'row',
    alignItems: 'flex-end',
  },
  myMsgTitle: {
    fontSize: 16,
    fontWeight: 'bold',
    color: '#153776',
    textAlign: 'right',
    paddingRight: '5%',
  },
  theirMsgTitle: {
    fontSize: 16,
    fontWeight: 'bold',
    color: '#153776',
    textAlign: 'left',
    paddingLeft: '5%',
  },
  myMessage: {
    backgroundColor: '#67b2da',
    padding: '4%',
    margin: '2%',
    marginLeft: '20%',
    borderRadius: 10,
  },
  theirMessage: {
    justifyContent: 'space-between',
    backgroundColor: '#dcdcdc',
    marginLeft: '5%',
    padding: '4%',
    margin: '2%',
    marginRight: '20%',
    borderRadius: 10,
  },
  myMsgText: {
    fontSize: 16,
    color: 'white',
  },
  theirMsgText: {
    fontSize: 16,
    color: 'black',
  },
  myTime: {
    fontSize: 12,
    color: 'white',
    textAlign: 'right',
  },
  theirTime: {
    fontSize: 12,
    color: '#6a6a6a',
  },
  avatar: {
    borderRadius: 50,
    width: 60,
    height: 60,
    marginLeft: '1%',
  },
  // system messages
  sysContainer: {
    justifyContent: 'center',
    marginHorizontal: '15%',
  },
  sysMessage: {
    fontSize: 16,
    backgroundColor: 'aquamarine',
    padding: '4%',
    margin: '2%',
    borderRadius: 10,
  },
  sysMsgText: {
    fontSize: 14,
    textAlign: 'center',
  },
});
