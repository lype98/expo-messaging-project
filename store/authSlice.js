import { createSlice } from '@reduxjs/toolkit';

const authSlice = createSlice({
  name: 'auth',
  initialState: {
    uid: '',
    email: '',
    username: '',
  },
  reducers: {
    login(state, action) {
      state.uid = action.payload.uid;
      state.email = action.payload.email;
      state.username = action.payload.username;
    },
    logOut(state) {
      for (let key in state) {
        state[key] = '';
      }
    },
  },
});

export const authActions = authSlice.actions;

export default authSlice;
