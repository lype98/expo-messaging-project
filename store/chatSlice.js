import { createSlice } from '@reduxjs/toolkit';

const chatSlice = createSlice({
  name: 'chat',
  initialState: {
    currentRoomID: null,
    currentRoomType: null,
    currentRoomUsers: null,
    refreshing: false,
  },
  reducers: {
    enterRoom(state, action) {
      state.currentRoomID = action.payload.roomID;
      state.currentRoomType = action.payload.chatType;
      state.currentRoomUsers = action.payload.users;
    },
    refresh(state) {
      state.refreshing = !state.refreshing;
    },
  },
});

export const chatActions = chatSlice.actions;

export default chatSlice;
