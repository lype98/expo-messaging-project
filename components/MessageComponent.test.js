import React from 'react';
import renderer from 'react-test-renderer';
import MessageComponent from './MessageComponent';
import { Provider } from 'react-redux';
import store from '../__mocks__/store/store-index';

const messageObj = {
  createdAt: {
    nanoseconds: 566000000,
    seconds: 1643977071,
  },
  key: 'zloV06XqRujF1RvmvvEx',
  message: 'C señoooooor',
  uid: 'kBp1Bvh2jIdPpnO8TfSiV59ZqtB2',
};

const dmTree = renderer.create(
  <Provider store={store}>
    <MessageComponent messageObj={messageObj} otherUser={'C señor'} chatType={'DM'} />
  </Provider>
);

const groupTree = renderer.create(
  <Provider store={store}>
    <MessageComponent messageObj={messageObj} otherUser={'Group Chat!'} chatType={'Group'} />
  </Provider>
);

it('renders with props', () => {
  expect(dmTree.toJSON()).toMatchSnapshot();
});

it('expects senderName to be C señor', () => {
  const msgTitle = dmTree.root.findByProps({ testID: 'msgTitle' }).props.children;
  expect(msgTitle).toBe('C señor');
});

it('expects group message sender to be C señor', () => {
  const msgTitle = groupTree.root.findByProps({ testID: 'msgTitle' }).props.children;
  expect(msgTitle).toBe('C señor');
});
