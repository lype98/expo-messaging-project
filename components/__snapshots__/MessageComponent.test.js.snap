// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with props 1`] = `
<View
  style={
    Object {
      "alignItems": "flex-end",
      "flexDirection": "row",
    }
  }
>
  <Image
    source={1}
    style={
      Object {
        "borderRadius": 50,
        "height": 60,
        "marginLeft": "1%",
        "width": 60,
      }
    }
  />
  <View>
    <Text
      style={
        Object {
          "color": "#153776",
          "fontSize": 16,
          "fontWeight": "bold",
          "paddingLeft": "5%",
          "textAlign": "left",
        }
      }
      testID="msgTitle"
    >
      C señor
    </Text>
    <View
      style={
        Object {
          "backgroundColor": "#dcdcdc",
          "borderRadius": 10,
          "justifyContent": "space-between",
          "margin": "2%",
          "marginLeft": "5%",
          "marginRight": "20%",
          "padding": "4%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "black",
            "fontSize": 16,
          }
        }
      >
        C señoooooor
      </Text>
      <Text
        style={
          Object {
            "color": "#6a6a6a",
            "fontSize": 12,
          }
        }
      >
        12:17 PM
      </Text>
    </View>
  </View>
</View>
`;
